automation/functions/delivery.py

```diff
@@ -0,0 +1,174 @@
+        log.error(msg, problem=problem)
+        return
+
+    delivery_volunteers = get_delivery_volunteers(ticket, member_table)
+    email = render_email_template(ticket, delivery_volunteers, inventory)
+    email.from_email = from_email
+    if reply_to:
+        email.add_cc(reply_to)
+        email.reply_to = reply_to
+
+    if not SEND_MAIL:
+        log.info(
+            "Would send email", to=email.get()["personalizations"][0]["to"]
+        )
+        return
+    try:
+        sendgrid_client.send(email)
+        log.info("Sent email")
+    except BadRequestsError as e:
+        msg = f"Error sending email: {e.status_code} {e.reason}"
+        log.error(msg, body=e.body, exc_info=True)
+        raise DeliveryEmailError(msg)
+
+
+def check_ready_to_send(ticket):
+    if ticket.status != "Assigned / In Progress":
+        return f"Ticket {ticket.ticket_id} is not assigned"
+    if not ticket.delivery_volunteer:
+        return f"Ticket {ticket.ticket_id} has no delivery volunteer"
+    if not ticket.request_name:
+        return f"Ticket {ticket.ticket_id} has no requester name"
+    if not ticket.address:
+        return f"Ticket {ticket.ticket_id} has no address"
+    if not ticket.phone_number:
+        return f"Ticket {ticket.ticket_id} has no phone number"
+    return None
+
+
+def get_delivery_volunteers(ticket, member_table):
+    delivery_volunteers = []
+    for r in ticket.delivery_volunteer:
+        try:
+            delivery_volunteers.append(member_table.get(r))
+        except HTTPError as e:
+            msg = f"Error finding delivery volunteer {r}: {e.response.text}"
+            log.error(msg, exc_info=True)
+            raise DeliveryEmailError(msg)
+
+    return delivery_volunteers
+
+
+def render_email_template(ticket, delivery_volunteers, inventory):
+    shopping_list = [
+        {
+            "name": item,
+            "category": inventory.category(item),
+            "quantity": inventory.quantity(item, ticket.household_size),
+            "unit": inventory.unit(item),
+        }
+        for item in ticket.food_options
+    ]
+    message = Mail(
+        to_emails=[f"{v.name} <{v.email}>" for v in delivery_volunteers],
+        subject=(
+            f"[Bed Stuy Strong] Delivery Instructions for {ticket.ticket_id}"
+        ),
+        html_content=render(
+            "delivery_email.html.jinja",
+            ticket=ticket,
+            delivery_volunteers=delivery_volunteers,
+            shopping_list=shopping_list,
+        ),
+    )
+    return message
+
+
+class Inventory:
+    def __init__(self, items_by_household_size_table):
+        self.categories = {}
+        self.quantities = {}
+        self.units = {}
+        for record in items_by_household_size_table.get_all(
+            '{Category} != "Children / Babies"'
+        ):
+            self.categories[record.item] = record.category
+            self.quantities[record.item] = [
+                0,
+                record.size_1,
```

Comment:
size_1-size_10 guaranteed to exist for all items on the househould_size_table?

Reply:
If the values are not guaranteed as integers, maybe default to `1` or the maximum integer from lower household sizes.

Reply:
Yes, they exist for all items (except diapers/formula, which we exclude by filtering out the children/babies category).

We maybe should be a little more defensive here so that if someone is editing that table (changing sizes or adding new items) it doesn't cause failures if the poll function runs at an inopportune moment. Let me think about how to defend better here.

Reply:
https://github.com/bedstuystrong/automation_prime/issues/60